src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -745,63 +746,204 @@ void StorageKafka2::dropReplica()
+
+    LOG_TRACE(
+        log,
+        "Locked topic partition: {}:{} at offset {} with intent size {}",
+        partition_to_lock.topic,
+        partition_to_lock.partition_id,
+        lock_info.committed_offset.value_or(0),
+        lock_info.intent_size.value_or(0));
+    locks.emplace(TopicPartition(partition_to_lock), std::move(lock_info));
+}
+
+
+void StorageKafka2::unlockTopicPartition(zkutil::ZooKeeper & keeper_to_use,  const TopicPartition& partition_to_unlock) {
+    const auto topic_partition_path = getTopicPartitionPath(partition_to_unlock);
+    const auto lock_file_path = String(topic_partition_path / lock_file_name);
+    Coordination::Error code = keeper_to_use.tryRemove(lock_file_path);
+    if (code == Coordination::Error::ZNONODE)
+        return;
+
+    LOG_TRACE(
+        log,
+        "Replica {} successfully unlocked topic partition {}:{}",
+        (*kafka_settings)[KafkaSetting::kafka_replica_name].value,
+        partition_to_unlock.topic,
+        partition_to_unlock.partition_id
+    );
+}
+
+/// If the number of locks on a replica is greater than it can hold, then we first release the partitions that we can no longer hold.
+/// Otherwise, we try to lock free partitions one by one.
+/// Also, each replica holds temporary locks. The replica removes these temporary locks at each new iteration of processing.
+std::optional<StorageKafka2::TopicPartitionLocks>
+StorageKafka2::lockTopicPartitions(zkutil::ZooKeeper & keeper_to_use, const TopicPartitions & topic_partitions)
+{
+    const auto already_locked_partitions = lookupReplicaState(keeper_to_use);
+    for (const auto & already_locks : already_locked_partitions) {
+        LOG_INFO(
+            log,
+            "Already locked topic partition {}:{}",
+            already_locks.topic,
+            already_locks.partition_id
+        );
+    }
 
-    if (const auto code = keeper_to_use.tryMulti(ops, responses); code != Coordination::Error::ZOK)
+    for (const auto & partition_to_unlock : ReplicaState.tmp_topics_assigned)
     {
-        if (code != Coordination::Error::ZNODEEXISTS)
-            zkutil::KeeperMultiException::check(code, ops, responses);
+        unlockTopicPartition(keeper_to_use, partition_to_unlock);
+    }
 
-        LOG_TRACE(log, "Couldn't create topic partitions locks because some of them already exists");
-        // Possible optimization: check the content of lock files, if we locked them, then we can clean them up and retry to lock them.
-        return std::nullopt;
+    size_t can_lock_partitions = static_cast<size_t>(std::ceil(static_cast<double>(topic_partitions.size()) / ActiveReplicaCount));
+    TopicPartitionLocks permanent_locks;
+    TopicPartitionLocks tmp_locks;
+
+    TopicPartitions available_topic_partitions;
+    available_topic_partitions.reserve(topic_partitions.size());
+    for (const auto & partition : topic_partitions)
+    {
+        if (!already_locked_partitions.contains(partition))
```

Comment:
Let's create an `std::unordered_set` to do lookups. I think in most of the cases it won't matter, but when dealing with a lot of topic partitions (some people uses ClickHouse to deal with serious amount of data) then an `std::vector` can be much slower.